### PR TITLE
Merge: merge server-python3 to master for version 1.0.0

### PR DESCRIPTION
**Pull request by rseuster**
probably more modifications coming, with this the RPM at least builds OK

Summary of changes on shoal-agent:
- Removed file relates to downloading geolitecity data, it needs manually update now
- Updated the verification process, if fails more than 4 test urls, then stop the verification; Removed the test for hostname 
- Updated the method to compare the domain of two ips
- Added local attribute to the shoal data to identify if a squid has the same domain as the client
- Removed the shoal data domain_acess, and use allow_verfication and global_access to handle the squid verification and set the access level
- Added a logger for shoal-server and use the logging level configable
- Added sleep interval for verification process
- Updated the UI, removed the column "verified", only use the "access level" column to show the access level; Update the "last received time" column number to be int
- Updated version to 1.0.0

### CODE REVIEW

File: shoal-server/setup.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
     try:
         from distutils.core import setup
     except:
-        print "Couldn't use either setuputils or distutils. Install one of those."
+        print ("Couldn't use either setuputils or distutils. Install one of those.")
```
Comment (MarcusEbert) [inline shoal-server/setup.py:11]: setuputils -> setuptools

File: shoal-server/shoal_server/config.py [hunk 2]
```diff
@@ -35,6 +35,7 @@
 amqp_client_key  = ''
 webpy_cache = False
 log_file = '/var/log/shoal_server.log'
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/config.py:37]: should be an option in the config file

File: shoal-server/shoal_server/config.py [hunk 3]
```diff
@@ -35,6 +35,7 @@
...
+logging_level = 'WARNING'
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/config.py:38]: should also be defined in the config file

File: shoal-server/shoal_server/config.py [hunk 4]
```diff
@@ -72,27 +73,27 @@
 elif exists(abspath(homedir + "/.shoal/shoal_server.conf")):
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/config.py:73]: should be second choice; local dir -> home dir -> system dir

File: shoal-server/shoal_server/shoal.py [hunk 5]
```diff
@@ -505,32 +514,17 @@ def on_message(self, unused_channel, basic_deliver, properties, body):
...
         if curr - time_sent > 10:
-            logging.error(
+            logger.error(
                 "Potential clock drift dectected: %s second descrepency on %s",
                 (curr-time_sent),
                 public_ip)
             drift_detected = True
         elif curr - time_sent < -10:
-            logging.error(
+            logger.error(
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/shoal.py:527]: "Recived message from %s  %s seconds in the future.",
￼                public_ip, (-1*(curr-time_sent)))

File: shoal-server/shoal_server/shoal.py [hunk 6]
```diff
@@ -14,10 +14,14 @@
 from shoal_server import config
 from shoal_server import utilities
 
+level = logging.getLevelName(config.logging_level)
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/shoal.py:17]: Here it gets the logging level which is fine (should come from config file), however in other parts of the code, the level is hardcoded to warning. Would be good to use everywhere the config value.

File: shoal-server/shoal_server/utilities.py [hunk 7]
```diff
@@ -114,40 +121,36 @@ def get_all_squids():
     return squids
 
 def lookupDomain(temp_ip):
-    name = socket.getfqdn(temp_ip)
-    if name == temp_ip:
-       	return None
-    domain = name.split('.')
-    return domain[-2]+'.'+domain[-1]
-
+#    name = socket.getfqdn(temp_ip)
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/utilities.py:124]: remove old unused code

**Review by MarcusEbert — CHANGES_REQUESTED** [review]

File: shoal-server/shoal_server/config.py (hunk 3 above)
Comment (colsond) [reply shoal-server/shoal_server/config.py:38]: This is the hard coded default that will later be overwritten by the content of shoal_server.conf if it is set there.

File: shoal-server/shoal_server/shoal.py (hunk 6 above)
Comment (colsond) [reply shoal-server/shoal_server/shoal.py:17]: This function converts the loglevel string (ie INFO) to a numeric value (ie 20) or vise versa, we then use this level variable below to set the log level. The only place i've seen warning hard coded is the default config value in config.py.

File: shoal-server/shoal_server/shoal.py [hunk 8]
```diff
@@ -597,17 +592,23 @@ def run(self):
...
-                # In an ideal scenario shoal would continuously try to verify those
-                # that are not verified. However since many squids use old agents
-                # its likely that they are not configured correctly and this would
-                # make the verification loop try and verify them over and over when
-                # they can't be. Instead they will be verified once every interval
-                # greatly reducing computing resource requirements for misconfigured agents.
-                if (current_time - squid.last_verified) >= INTERVAL:
-                    utilities.verify(squid)
-                    squid.last_verified = time()
+            for squid_key in list(web.shoal):
```
Comment (colsond) [inline shoal-server/shoal_server/shoal.py:595]: Make a list of the shoal keys to loop over so you dont get data manipulation errors.

File: shoal-server/shoal_server/utilities.py [hunk 9]
```diff
@@ -21,7 +23,9 @@
 GEOLITE_UPDATE = config.geolitecity_update
 GEODOMAIN_DB = os.path.join(config.geodomain_path, "GeoIP2-Domain.mmdb")
 
+level = logging.getLevelName(config.logging_level)
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:26]: this statement is probably superfluous, you can instead pass config.logging_level directly to setLevel.

File: shoal-server/shoal_server/utilities.py [hunk 10]
```diff
@@ -60,7 +64,7 @@ def get_nearest_verified_squids(ip, count=10):
     """
     request_data = get_geolocation(ip)
     if not request_data:
-        print "no geolocation"
+        print ("no geolocation")
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:67]: update prints to log statements. This particular statement could be level debug. Maybe update the message to include IP ("No geolocation for %s" % ip).

File: shoal-server/shoal_server/utilities.py [hunk 11]
```diff
@@ -114,40 +121,36 @@ def get_all_squids():
...
+#    domain = name.split('.')
+#    return domain[-2]+'.'+domain[-1]
+    try:
+        findDomain = subprocess.Popen('dig +nocomments  -x' + temp_ip + 'soa', shell=True,stdout=subprocess.PIPE,stderr=subprocess.STDOUT)
+        result = findDomain.communicate()[0].decode('utf-8').splitlines()
+        for entry in result:
+            if entry != '' and not entry.startswith(';'):
+                return entry.split()[5]
+        return None
+    except:
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:136]: Catch and log this exception so its easier for an administrator to troubleshoot.

File: shoal-server/shoal_server/utilities.py [hunk 12]
```diff
@@ -114,40 +121,36 @@ def get_all_squids():
...
+            return True
+        else:
             return False
-    else:
-        logging.error("No geoDomain database file detected. Add the database file "
-                      "to shoal-server/static/db before installation and ensure the path in "
-                      "the config file is correct")
+    except Exception as exc:
+        logger.error(exc)
+        logger.error("Could not compare the domain for the two ips")
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:153]: Add IPs to log statement requestip: %s and squid_ip: %s

File: shoal-server/shoal_server/utilities.py [hunk 13]
```diff
@@ -263,74 +243,53 @@ def _is_available(squid):
...
+                        ip, targeturl, badflags, len(paths))
+            except:
+                # note that this would catch any RE errors aswell but they are
+                # specified in the config and all fit the pattern.
+                badpaths = badpaths + 1
+                #logging.error(sys.exc_info()[1])
+                logger.error("Timeout or proxy error on %s repo. Currently %s out of %s repos failing", targeturl, badpaths, len(paths))
+            finally:
+                #Keep going
+                logger.info("Next...")
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:275]: this can be a debug level message

File: shoal-server/shoal_server/utilities.py [hunk 14]
```diff
@@ -263,74 +243,53 @@ def _is_available(squid):
...
-        if squid.domain_access and squid.global_access:
+        # todo: rewrite the error cases when squid doesn't have a domain_access attribute
+        if squid.global_access:
             squid.error = "Configuration conflict detected! %s is configured " \
                           "for both Local Access Only and Global Access."  % (hostname)
 
-        if squid.domain_access and not squid.global_access:
+        if not squid.global_access:
             squid.error = "Squid is configured for Local Access Only. Cannot verify %s" % (hostname)
         else:
```
Comment (colsond) [inline shoal-server/shoal_server/utilities.py:291]: Nothing will fall into this else statement so it can be removed.

**Review by colsond — CHANGES_REQUESTED** [review]

File: shoal-server/shoal_server/shoal.py [hunk 15]
```diff
@@ -76,7 +81,7 @@ def __init__(self, shoal):
 
         # check if geolitecity database needs updating
         if utilities.check_geolitecity_need_update():
-            utilities.download_geolitecity()
+            logger.warning('GeoLiteCity database needs manually updating.')
```
[the remark this answers is not in the document]
Comment (MarcusEbert) [reply shoal-server/shoal_server/shoal.py:84]: GeoLiteCity db is still used to make the match based on distance. This was the one we thought about converting to mysql.

File: shoal-server/shoal_server/utilities.py [hunk 16]
```diff
@@ -21,7 +23,9 @@
 GEOLITE_UPDATE = config.geolitecity_update
 GEODOMAIN_DB = os.path.join(config.geodomain_path, "GeoIP2-Domain.mmdb")
```
Comment (MarcusEbert) [inline shoal-server/shoal_server/utilities.py:24]: I believe the domain db is no longer used and replaced by the dig calls to determine the domain of an IP.

**Review by MarcusEbert — CHANGES_REQUESTED** [review]

File: shoal-server/conf/shoal_wsgi.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
+import logging
```
[the remark this answers is not in the document]
Comment (MCatherine1994) [reply shoal-server/conf/shoal_wsgi.py:1]: This file appears at two locations, "shoal/shoal-server/scripts/shoal_wsgi.py" and "shoal/shoal-server/conf/scripts/shoal_wsgi.py" and have the same content. In the setup file, it includes "conf/scripts/shoal_wsgi.py", and I didn't see any usage of the other one.

File: shoal-server/setup.py [hunk 18]
```diff
@@ -34,7 +34,7 @@
       packages=['shoal_server'],
       scripts=["shoal-server"],
       data_files=[('share/shoal-server', ['conf/shoal_server.conf', 'conf/shoal.conf','conf/shoal-server.logrotate']),
-                       ('share/shoal-server/scripts', ['conf/scripts/shoal_wsgi.py','conf/scripts/setup_files.py']),
+                       ('share/shoal-server/scripts', ['conf/scripts/shoal_wsgi.py']),
```
Comment (MCatherine1994) [inline shoal-server/setup.py:37]: This is where the shoal_wsgi.py has been included. Didn't see the other shoal_wsgi.py file has been used anywhere

File: shoal-server/shoal_server/config.py [hunk 19]
```diff
@@ -1,6 +1,6 @@
 from os.path import join, expanduser, exists, abspath
 import sys
-import ConfigParser
+import configparser
```
[the remark this answers is not in the document]
Comment (MCatherine1994) [reply shoal-server/shoal_server/config.py:3]: we only confirmed the support of python3 for shoal-server

File: shoal-server/shoal_server/utilities.py (hunk 9 above)
Comment (MCatherine1994) [reply shoal-server/shoal_server/utilities.py:25]: tested and updated

File: shoal-server/shoal_server/utilities.py (hunk 16 above)
Comment (MCatherine1994) [reply shoal-server/shoal_server/utilities.py]: removed

File: shoal-server/shoal_server/utilities.py (hunk 14 above)
Comment (MCatherine1994) [reply shoal-server/shoal_server/utilities.py]: Removed the if-else condition to check squid.global_acess, as this value is initialized as false, and will only become true when it passes the verification tests. So when at this step, its value is always false. Updated the to just have squid.error = "%s/%s URLs have proxy errors and %s/%s URLs are unreachable. Squid is configured for Local Access Only. Cannot verify %s" % (badpaths, len(paths), badflags, len(paths), hostname)"

File: shoal-server/shoal_server/config.py (hunk 4 above)
Comment (colsond) [reply shoal-server/shoal_server/config.py:71]: For the server all config file locations can be removed except for the one in /etc/shoal/

**Review by colsond — CHANGES_REQUESTED** [review]

**Review by MarcusEbert — APPROVED** [review]

**Review by colsond — APPROVED** [review]

**Pull request merged**